decoder.go

```diff
@@ -454,39 +456,79 @@ func (d *decoder) pointer(name string, value reflect.Value, def string) error {
 	return d.unmarshal(name, value.Elem(), def)
 }
 
-// Sets value to a channel type.
-func (d *decoder) channel(key string, value reflect.Value, def string) error {
-	return d.textUnmarshaller(key, value, def)
-}
-
-// Sets value to a function type.
-func (d *decoder) function(key string, value reflect.Value, def string) error {
-	return d.textUnmarshaller(key, value, def)
-}
+// JSON encoder will fail to serialize maps that don't have strings as keys
+// so we are going to stringify them manually.
+func jsonMap(v interface{}) interface{} {
+	if reflect.TypeOf(v).Kind() == reflect.Map {
+		tmp := make(map[string]interface{})
+		rv := reflect.ValueOf(v)
+		for _, key := range rv.MapKeys() {
+			tmp[fmt.Sprint(key.Interface())] = jsonMap(rv.MapIndex(key).Interface())
```

Comment:
Unfortunately we can't do this: a key could be of any type, e.g. integer. The reason we have to stringify map keys is because JSON marshaller supports only maps with string keys.
Another option would be to return an error, but I want to do our best first and try to populate with a new map instead of simply returning an error.